src/middlewared/middlewared/plugins/smb.py

```diff
@@ -110,3 +112,255 @@ class Config:
+    async def do_delete(self, id):
+        return await self.middleware.call(
+            'datastore.delete', self._config.datastore, id)
+
+    @private
+    async def clean(self, data, schema_name, verrors, id=None):
+        data['name'] = await self.name_exists(data, schema_name, verrors, id)
+
+    @private
+    async def validate(self, data, schema_name, verrors, old=None):
+        home_result = await self.home_exists(
+            data['home'], schema_name, verrors, old)
+
+        if home_result:
+            verrors.add(f'{schema_name}.home',
+                        'Only one share is allowed to be a home share.')
+        elif not home_result and not data['path']:
+            verrors.add(f'{schema_name}.path', 'This field is required.')
+
+    @private
+    async def home_exists(self, home, schema_name, verrors, old=None):
+        home_filters = [('home', '=', True)]
+        home_result = None
+
+        if home:
+            if old and old['id'] is not None:
+                id = old['id']
+
+                if not old['home']:
+                    home_filters.append(('id', '!=', id))
+                    # The user already had this set as the home share
+                    home_result = await self.middleware.call(
+                        'datastore.query', self._config.datastore,
+                        home_filters, {'prefix': self._config.datastore_prefix})
+
+        return home_result
+
+    async def name_exists(self, data, schema_name, verrors, id=None):
+        name = data['name']
+        path = data['path']
+        name_filters = [('name', '=', name)]
+        path_filters = [('path', '=', path)]
+
+        if path and not name:
+            name = path.rsplit('/', 1)[-1]
+
+        if id is not None:
+            name_filters.append(('id', '!=', id))
+            path_filters.append(('id', '!=', id))
+
+        name_result = await self.middleware.call(
+            'datastore.query', self._config.datastore,
+            name_filters,
+            {'prefix': self._config.datastore_prefix})
+        path_result = await self.middleware.call(
+            'datastore.query', self._config.datastore,
+            path_filters,
+            {'prefix': self._config.datastore_prefix})
+
+        if name_result:
+            verrors.add(f'{schema_name}.name',
+                        'A share with this name already exists.')
+
+        if path_result:
+            verrors.add(f'{schema_name}.path',
+                        'A share with this path already exists.')
+
+        return name
+
+    @private
+    async def extend(self, data):
+        data['hostsallow'] = data['hostsallow'].split()
+        data['hostsdeny'] = data['hostsdeny'].split()
+
+        return data
+
+    @private
+    async def compress(self, data):
+        data['hostsallow'] = ' '.join(data['hostsallow'])
+        data['hostsdeny'] = ' '.join(data['hostsdeny'])
+
+        return data
+
+    @private
+    async def done(self, default_perms, path):
```

Comment:
Applying default perms is all we do here, how about naming the method accordingly?